src/Util/SmartPaginatorFactory.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the Sonata Project package.
+ *
+ * (c) Thomas Rabaix <thomas.rabaix@sonata-project.org>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sonata\DoctrineORMAdminBundle\Util;
+
+use Doctrine\ORM\Tools\Pagination\CountWalker;
+use Doctrine\ORM\Tools\Pagination\Paginator;
+use Sonata\DoctrineORMAdminBundle\Datagrid\ProxyQuery;
+use Sonata\DoctrineORMAdminBundle\Datagrid\ProxyQueryInterface;
+
+/**
+ * @internal
+ */
+final class SmartPaginatorFactory
+{
+    /**
+     * NEXT_MAJOR: Replace ProxyQuery by ProxyQueryInterface.
+     *
+     * @param array<string, mixed> $hints
+     */
+    public static function create(ProxyQuery $proxyQuery, array $hints = []): Paginator
+    {
+        $queryBuilder = $proxyQuery->getQueryBuilder();
+
+        $identifierFieldNames = $queryBuilder
+            ->getEntityManager()
+            ->getClassMetadata(current($queryBuilder->getRootEntities()))
+            ->getIdentifierFieldNames();
+
+        $hasSingleIdentifierName = 1 === \count($identifierFieldNames);
+        $hasJoins = \count($queryBuilder->getDQLPart('join')) > 0;
+
+        $query = $proxyQuery->getDoctrineQuery();
+
+        if (!$hasJoins) {
+            $query->setHint(CountWalker::HINT_DISTINCT, false);
+        }
+
+        foreach ($hints as $name => $value) {
+            $query->setHint($name, $value);
+        }
+
+        // Paginator with fetchJoinCollection doesn't work with composite primary keys
+        // https://github.com/doctrine/orm/issues/2910
+        // To stay safe fetch join only when we have single primary key and joins
+        $paginator = new Paginator($query, $hasSingleIdentifierName && $hasJoins);
```

Comment:
Do we plan to add an option in order to let the user decide if `fetchJoin` should be true ?
You might want to set `fetchJoin = false` even if there are join, because you're not using fetch join ; no ?

If so, the config will be pass in the constructor of the Factory and this method won't be able to be static. Should we anticipate this now ?

Reply:
TBH I don't know, I created `SmartPaginatorFactory` as internal to just extract that code, so it could be changed the way we want. I would wait to do that when these options are created.